pkg/utils/subst/subst.go

```diff
@@ -36,110 +42,156 @@ func ParseFile(file string, fss ...vfs.FileSystem) (SubstitutionTarget, error) {
 	return &fi, nil
 }
 
 type fileinfo struct {
-	content *ast.File
+	content *yqlib.CandidateNode
 	json    bool
 }
 
 func (f *fileinfo) Content() ([]byte, error) {
-	data := []byte(f.content.String())
-
+	var enc yqlib.Encoder
 	if f.json {
-		// TODO: the package seems to keep the file type json/yaml, but I'm not sure
-		var err error
-		data, err = yaml.YAMLToJSON(data)
-		if err != nil {
-			return nil, errors.Wrapf(err, "cannot marshal json")
-		}
+		prfs := yqlib.NewDefaultJsonPreferences()
+		prfs.ColorsEnabled = false
+		enc = yqlib.NewJSONEncoder(prfs)
+	} else {
+		prfs := yqlib.NewDefaultYamlPreferences()
+		enc = yqlib.NewYamlEncoder(prfs)
 	}
-	return data, nil
-}
 
-func (f *fileinfo) SubstituteByData(path string, value []byte) error {
-	var m interface{}
-	err := runtime.DefaultYAMLEncoding.Unmarshal(value, &m)
-	if err != nil {
-		return err
-	}
-	if f.json {
-		value, err = runtime.DefaultJSONEncoding.Marshal(m)
+	buf := bytes.NewBuffer([]byte{})
+	pw := yqlib.NewSinglePrinterWriter(buf)
+	p := yqlib.NewPrinter(enc, pw)
+	inptLst := list.New()
+	inptLst.PushBack(f.content)
+
+	if err := p.PrintResults(inptLst); err == nil {
+		return buf.Bytes(), nil
 	} else {
-		value, err = runtime.DefaultYAMLEncoding.Marshal(m)
+		return nil, err
 	}
-	if err != nil {
-		return err
-	}
-	return f.substituteByData(path, value)
 }
 
-func (f *fileinfo) substituteByData(path string, value []byte) error {
-	file, err := parser.ParseBytes(value, 0)
+func (f *fileinfo) SubstituteByData(path string, value []byte) error {
+	m := &yaml.Node{}
+	err := yaml.Unmarshal(value, m)
 	if err != nil {
-		return errors.Wrapf(err, "cannot unmarshal value")
+		return err
 	}
 
-	p, err := yaml.PathString("$." + path)
-	if err != nil {
-		return errors.Wrapf(err, "invalid substitution path")
+	if !f.json {
+		var replaceFlowStyle func(*yaml.Node)
+		replaceFlowStyle = func(nd *yaml.Node) {
+			if nd.Style == yaml.FlowStyle {
+				nd.Style = yaml.LiteralStyle
+			}
+			for _, chld := range nd.Content {
+				replaceFlowStyle(chld)
+			}
+		}
+		replaceFlowStyle(m)
 	}
 
-	return p.ReplaceWithFile(f.content, file)
+	nd := &yqlib.CandidateNode{}
+	nd.SetDocument(0)
+	nd.SetFilename("value")
+	nd.SetFileIndex(0)
+
+	if err = nd.UnmarshalYAML(m.Content[0], map[string]*yqlib.CandidateNode{}); err != nil {
+		return err
+	}
+	return f.substituteByValue(path, nd)
 }
 
 func (f *fileinfo) SubstituteByValue(path string, value interface{}) error {
-	var (
-		err  error
-		data []byte
-	)
+	var mrshl func(interface{}) ([]byte, error)
+
 	if f.json {
-		data, err = runtime.DefaultJSONEncoding.Marshal(value)
+		mrshl = runtime.DefaultJSONEncoding.Marshal
 	} else {
-		data, err = runtime.DefaultYAMLEncoding.Marshal(value)
+		mrshl = runtime.DefaultYAMLEncoding.Marshal
 	}
+
+	if bval, err := mrshl(value); err != nil {
+		return err
+	} else {
+		return f.SubstituteByData(path, bval)
+	}
+}
+
+func (f *fileinfo) substituteByValue(path string, value *yqlib.CandidateNode) error {
```

Comment:
Is there a reason for using all these substitution parsings instead of a `NewAllAtOnceEvaluator` that produces a `[]byte` result given a printer?

Reply:
- As mentioned above, I wouldn't want to switch the code to hang on to a byte array instead of the AST
- The combo of NewAllAtOnceEvaluator the returned Evaluator interface don't let me set the Context and that is how we pass the substitution value to yqlib

Reply:
👍